---
rule_id: {DOMAIN}-{SEVERITY}{NUMBER}
domain: {DOMAIN}
severity: {Critical|High|Medium|Low}
priority: {P|H|M|L}
number: {001-999}
tags: [tag1, tag2]
auto_fixable: {true|false}
related_rules: [{RULE-ID}]
---

# Rule: {DOMAIN}-{SEVERITY}{NUMBER} {#domain-severity-number}

> **{ğŸ”´|ğŸŸ |ğŸŸ¡|ğŸŸ¢} {SEVERITY}** | **Auto-Fix:** {âœ…|âŒ}

## ğŸ“‹ Pattern

{One-sentence description}

## â“ Why This Matters

{Brief explanation of why this rule exists}

## ğŸ” Detection

```bash
{detection command}
```

## âœ… Fix Instructions

{Quick fix description}

## ğŸ“ Examples

### âŒ WRONG
```{language}
{code showing violation}
```

### âœ… CORRECT
```{language}
{code showing correct pattern}
```

## ğŸ”— References

- {Reference link or file path}

## ğŸ”€ Related Rules

- [{RULE-ID}](#{rule-id})

---
